packages/core/src/routes/experience/classes/interaction-session.ts

```diff
@@ -0,0 +1,170 @@
+  public setInteractionEvent(event: InteractionEvent) {
+    // TODO: conflict event check (e.g. reset password session can't be used for sign in)
+    this.interactionEvent = event;
+  }
+
+  /** Set the verified accountId of the current interaction session from  the verification record */
+  public identifyUser(verificationRecord: VerificationRecord) {
+    // Throws an 404 error if the user is not found by the given verification record
+    assertThat(
+      verificationRecord.verifiedUserId,
+      new RequestError(
+        {
+          code: 'user.user_not_exist',
+          status: 404,
+        },
+        {
+          identifier: verificationRecord.identifier.value,
+        }
+      )
+    );
+
+    this.accountId = verificationRecord.verifiedUserId;
+  }
+
+  /**
+   * Append a new verification record to the current interaction session.
+   * @remark If a record with the same type already exists, it will be replaced.
+   */
+  public appendVerificationRecord(record: VerificationRecord) {
+    const { type } = record;
+
+    const existingRecord = this.getVerificationRecordByType(type);
+
+    if (existingRecord) {
+      this.verificationRecords.delete(existingRecord);
+    }
+
+    this.verificationRecords.add(record);
+  }
+
+  public getVerificationRecordById(verificationId: string) {
+    return this.verificationRecordsArray.find((record) => record.id === verificationId);
+  }
+
+  /** Save the current interaction session result */
+  public async save() {
+    // The "mergeWithLastSubmission" will only merge current request's interaction results,
+    // manually merge with previous interaction results
+    // refer to: https://github.com/panva/node-oidc-provider/blob/c243bf6b6663c41ff3e75c09b95fb978eba87381/lib/actions/authorization/interactions.js#L106
+
+    const { provider } = this.tenant;
+    const details = await provider.interactionDetails(this.ctx.req, this.ctx.res);
+
+    await provider.interactionResult(
+      this.ctx.req,
+      this.ctx.res,
+      { ...details.result, ...this.toJson() },
+      { mergeWithLastSubmission: true }
+    );
+  }
+
+  /** Submit the current interaction session result to the OIDC provider and clear the session */
+  public async submit() {
+    // TODO: refine the error code
+    assertThat(this.accountId, 'session.verification_session_not_found');
+
+    const { provider } = this.tenant;
+
+    const redirectTo = await provider.interactionResult(this.ctx.req, this.ctx.res, {
+      login: { accountId: this.accountId },
+    });
+
+    this.ctx.body = { redirectTo };
+  }
+
+  private get verificationRecordsArray() {
+    return Array.from(this.verificationRecords);
+  }
+
+  private getVerificationRecordByType(type: VerificationType) {
+    return this.verificationRecordsArray.find((record) => record.type === type);
+  }
+
+  /** Convert the current interaction session to JSON, so that it can be stored as the OIDC provider interaction result */
+  private toJson() {
```

Comment:
maybe `toJsonObject`? since it's hard to know if it is a JSON string or an object.

Reply:
I doubt it. As JSON is an object, `toJson` should not return a JSON string?